Add blog post about avoiding the default address when using kubeadm (#28331)
kubernetes/website: content/en/blog/_posts/2021-06-08-Kubeadm-Avoid-Default-Address.md

---
layout: blog
title: "Avoiding the Default Address when Using kubeadm to Create a Cluster"
date: YYYY-MM-DD
slug: avoiding-the-default-address-when-using-kubeadm-to-create-a-cluster
---

**Authors:** Mike Spreitzer

When using `kubeadm` to create a cluster, the default behavior is to
take the default address of each node as the address to use in every
sort of Kubernetes configuration --- and sometimes this is entirely
wrong.  I am concerning myself here only with IPv4; extension to IPv6
and dual-stack should be follow the same lines.  The default address
of a node is the address of the network interface that appears in the
default route.  In some use cases, this default address is not
reachable from other nodes and so _must not_ be used in any part of
Kuberntes configuration.  In a recent adventure of mine, this arose
because each node is a VirtualBox VM that uses what VirtualBox calls a
"NAT network adapter" to get connectivity to the outside world.  You

> may have other use cases, perhaps involving containers.

Reviewer: This'd work better with a bit of a back story / scene setting (IMO). You needed to deploy a hybrid cluster that runs partly on bare metal with interesting constraints. Your cluster will run NetworkPolicy smoketests so doesn't look like a typical CNI. You just can't get enough VirtualBox.

Tell the reader about your recent adventures (in brief) and you will hook them to read on.

Author: The real story is really very simple and plain.  I wanted to run a quick experiment using VirtualBox, and I went about it in the most direct way.  If I were willing to put more work into it, I could have created a more elaborate set of VMs, adding a gateway that all the cluster VMs use; that would be a more normal networking topology, and the default address of each cluster node would naturally be distinct.  There are also other ways to not have the problem.  I suspect there are also other ways to have the problem, but can not say I actually ran those scenarios and have not put a lot of thought into building a bullet-proof case for them.  What I can say is that I ran into a problem that arises when you take the most simple and direct approach.

Author: I took a stab at this.

In this post I will look at how to make all of the changes needed for
a single-control-node cluster.  I have not yet attempted multiple
control nodes, and the solution will need further work for that case.
The needed results are getting the desired address to appear in the
following places.

- The IP address that the kube-apiserver puts in its Endoints object.
- The Subject Alternative Name list of the x509 certificate generated
  for the kube-apiserver.
- The IP address that appears in the `kubeadm join` commands that
  `kubeadm` outputs.
- The InternalIP type of address that each kubelet puts into its
  Node's status.

This post describes work done with version 1.21.1 of kubeadm and

> kubernetes.

Reviewer: ```suggestion
Kubernetes.
```

You may also like to read a related blog post, [Kubernetes Setup Using
Ansible and

> Vagrant](https://kubernetes.io/blog/2019/03/15/kubernetes-setup-using-ansible-and-vagrant/).

Reviewer: ```suggestion
Vagrant](/blog/2019/03/15/kubernetes-setup-using-ansible-and-vagrant/).
```

(unless I take extra steps).  This is relevant because in Ubuntu
20.04, the default DHCP client is `systemd.networkd`, and its default
configuration causes it to send an explicit "DHCP client identifier"
(in addition to the client MAC address) in each request and that
identifier is derived from `/etc/machine-id` and _not_ the relevant
network interface's MAC address.  When an explicit client ID is
supplied, the DHCP server uses only that --- not also the client MAC
address.  Howevever, rather than having to tweak each derived VM, I
tweaked the common snapshot to make the DHCP client eschew that
explicit client ID --- leaving the DHCP server to look only at the
client's MAC address.  The relevant bit of `systemd.networkd`
configuration is normally driven by a higher level tool called
`netplan`.  I tweaked its configuration by creating a drop-in file
named `/etc/netplan/10-enp0s8.yaml` (because `enp0s8` is the name of
the network interface that I want to get a distinct IPv4 address on
each node) with the following contents.

```yaml
network:
  ethernets:
    enp0s8:
      dhcp4: true
      dhcp-identifier: mac
  version: 2
```

## The kube-apiserver's address in its Endpoints object

> 

Reviewer: Maybe remind readers what an Endpoints is (I think the kube-apiserver creates a Service in the default namespace, and I'm not 100% sure on how the related Endpoints / EndpointSlices come into being) - it's a niche topic for some readers.

Author: Added.

The default for this can be overridden by passing
`--advertise-address=${desired_address}` on the kube-apiservers's
command line.  You can get `kubeadm init` to do that either with a
command line flag or a bit of configuration file content.  The command
line flag would be `--apiserver-advertise-address=192.168.56.101`,
assuming the desired address for the control node is 192.168.56.101.
I will use that example address throughout this post.

I used a configuration file, named `ka-init.conf`.  The following
content will get this job done.

```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
apiServer:
  extraArgs:
    advertise-address: 192.168.56.101
```

> ## The Subject Alternative Name list in the kube-apiserver's certificate

Reviewer: I'd actually write:
```suggestion
## The subjectAltName list in the kube-apiserver's certificate
```

Author: Why?  I think "Subject Alternative Name" is the canonical jargon.  That is what `openssl x509 -text` prints, and I find it in many articles about these certificates.

This job can be done with either a command line flag or a bit of
config file contents.  The command line flag is
`--apiserver-cert-extra-sans=192.168.56.101`.  The config file content
sets the `apiServer.certSANS` of a `ClusterConfiguration` to hold the
desired extra entries, such as `IP:192.168.56.101`.  Because the

> Subject Alternative Name list is a list, it is not functionally

Reviewer: ```suggestion
certificate's subjectAltName is a list, it is not functionally
```

critical to make that list exclude the undesired default address.

This job together with replacing the address
that appears in the generated `kubeadm join` command can be done with
a different command line flag or bit of configuration content.

## The IP address in generated `kubeadm join` commands

This is overridden with a command line flag or bit of config file
contents.  This same bit of kubeadm configuration also gets the

> desired address into the kube-apisever's x509 certificate's Subject
> Alternative Name list.

Reviewer: nit:
```suggestion
desired address into the kube-apisever's X.509 certificate's
subjectAltName list.
```

The command line flag is `--control-plane-endpoint=192.168.56.101`.

The config file contents sets the `controlPlaneEndpoint` of a
`ClusterConfiguration`.  Following is the cumulative config file
content so far.

```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
apiServer:
  extraArgs:
    advertise-address: 192.168.56.101
controlPlaneEndpoint: 192.168.56.101
```

## Each node's InternalIP type of address

The `kubeadm` tool takes no responsibility for configuring this.
However, it is part of the problem at hand.  I solved it by looking
into the `systemd` configuration for the kubelet and discovering that:
- the kubelet command line includes `$KUBELET_EXTRA_ARGS`,
- in a vanilla installation, nothing sets that variable,
- the file `/etc/default/kubelet`, if it exists, is `source`d as part
  of preparing the kubelet command line,
- in a vanilla installation, that file does not exist.

So I was able to use a simple solution that does not conflict with any
maintained configuration.  On each node I create the file named

> `/etc/default/kubelet` and give it content like the following.

Reviewer: ```suggestion
`/etc/default/kubelet` and give it content like the following:
```

```shell
KUBELET_EXTRA_ARGS=--node-ip=192.168.56.101
```

Note: the kubelet also accepts a command line flag named `--address`,
but that one does not get this job done.

**Note well:** the proper content for this file is different on every

> node.

Reviewer: Maybe coalesce these somehow?

Author: Perhaps something like the following?
> **NOTE**
> 1. The kubelet also accepts a command line flag named `--address`, but that one does not get this job done.
> 2. The proper content for this file is different on every node.

Author: replaced

## The grand finale

With the above in place, I started creating my two-node cluster by
executing

```shell
kubeadm init --config ka-init.conf
```

on the one control node and noting the `kubeadm join` command that was
printed at the end.

Next I used `kubectl` to look around an verify that everything looks
good.  The kube-apiserver's Endpoints object looks good.

```shell

> root@init1:~# kubectl get Endpoints -A

Reviewer: If including the prompt, don't request highlighting as shell (the prompt is not a valid command). Same for including output.

Author: Is there something better than "shell" to use in this case, or is it best to say nothing here?

Author: For now I simply struck `shell`.

> root@init1:~# kubectl get Endpoints -A

Reviewer: Optionally: use `kubectl get endpointslice --all-namespaces` instead. Still works OK; bit more modern, long options are self-documenting.

Author: Yes, long options are better in a blog post.
Also, the display format for EndpointSlice objects is more efficient;
`kubectl get Endpoints` produces truncated output, but not so when querying EndpointSlice objects.

NAMESPACE     NAME         ENDPOINTS             AGE
default       kubernetes   192.168.56.101:6443   30s
kube-system   kube-dns     <none>                14s
```

> The control node has the right address but is not ready.

Reviewer: ```suggestion
This control plane node has the right IPv4 address but is not ready
```
or
```suggestion
This shows that the Kubernetes control plane node is using the right IPv4 address. However,
there is more to do.
```

Investigation (not shown here) shows that it is because there is no

> network plugin yet.

Reviewer: ```suggestion
network plugin installed yet. Checking node status reveals:
```
?

kube-system   coredns-558bd4d5db-xjdlp        1/1     Running   0          2m38s   10.244.0.2       init1   <none>           <none>
kube-system   etcd-init1                      1/1     Running   0          2m50s   192.168.56.101   init1   <none>           <none>
kube-system   kube-apiserver-init1            1/1     Running   0          2m50s   192.168.56.101   init1   <none>           <none>
kube-system   kube-controller-manager-init1   1/1     Running   0          2m50s   192.168.56.101   init1   <none>           <none>
kube-system   kube-flannel-ds-hqtfh           1/1     Running   0          2m2s    192.168.56.101   init1   <none>           <none>
kube-system   kube-flannel-ds-nrn6n           1/1     Running   0          15s     192.168.56.102   join1   <none>           <none>
kube-system   kube-proxy-bn8s4                1/1     Running   0          2m38s   192.168.56.101   init1   <none>           <none>
kube-system   kube-proxy-wxvpg                1/1     Running   0          15s     192.168.56.102   join1   <none>           <none>
kube-system   kube-scheduler-init1            1/1     Running   0          2m50s   192.168.56.101   init1   <none>           <none>

> ```

Reviewer: How would you wrap this up?

Maybe a paragraph to say that you hope it's useful to show how kubeadm troubleshooting works, and also in case someone is facing a similar problem?

Author: Added a draft.

---
layout: blog
title: "Avoiding the Default Address when Using kubeadm to Create a Cluster"
date: 2021-06-08
slug: avoiding-the-default-address-when-using-kubeadm-to-create-a-cluster
---

**Authors:** Mike Spreitzer

When using `kubeadm` to create a cluster, the default behavior is to
take the default address of each node as the address to use in every
sort of Kubernetes configuration --- and sometimes this is entirely
wrong.  I am concerning myself here only with IPv4; extension to IPv6
and dual-stack should be follow the same lines.  The default address
of a node is the address of the network interface that appears in the
default route.  In some use cases, this default address is not
reachable from other nodes and so _must not_ be used in any part of

> Kuberntes configuration.  I recently had this happen to me, when I

Author: True, it is a pervasive issue.

> Kuberntes configuration.  I recently had this happen to me, when I

Reviewer 2: it is a fallback mechanism, if you need more complex routing you can always configure each component, I think that it is a good solution honestly

tried to do a quick experiment using VirtualBox VMs created in the
most direct and simple way using the VirtualBox GUI.  I gave each VM
two network interfaces: one using a "NAT adapter" to support
connectivity with the external world, and one using a "Host-only
adapter" for connectivity between the VMs.  The NAT adapter interfaces
all get the same IP address, 10.0.2.15, and so do not provide
connectivity between the VMs _but_ appear in the default route.  You

> may have other use cases, perhaps involving containers.

Author: I am not sure I follow what you mean about the VBox NAT adapter, but I am with you on your main point.  If there is an easier way to configure all the kube stuff to use a different interface/address, I would like to know and see it well documented and easy to find.

What you showed is what I also see when using the VBox NAT adapter.  And I see the same thing in every VM, even when they do not share the same /etc/machine-id contents.  I mean, all the VMs using the NAT adapter get the IP address 10.0.2.15 on that adapter.  That means they cannot communicate with each other using that address.  But that is the address in the default route, so it is the one all the kube stuff wants to use by default.

> may have other use cases, perhaps involving containers.

Reviewer 3: > I am not sure I follow what you mean about the VBox NAT adapter, but I am with you on your main point. If there is an easier way to configure all the kube stuff to use a different interface/address, I would like to know and see it well documented and easy to find.

AFAIK, there is no easier way to auto-configure IPs from a different interface.
i don't know if the k8s project maintains more detailed information about this. likely not.

> What you showed is what I also see when using the VBox NAT adapter. And I see the same thing in every VM, even when they do not share the same /etc/machine-id contents. I mean, all the VMs using the NAT adapter get the IP address 10.0.2.15 on that adapter. That means they cannot communicate with each other using that address. But that is the address in the default route, so it is the one all the kube stuff wants to use by default.

frankly, i'm not an expert around the different vbox network adapter options.
https://www.brianlinkletter.com/2016/07/how-to-use-virtualbox-to-emulate-a-network/
i know i can use "bridged" which maps VM addresses on the host network, which is what i'd like for my use cases and is suits the k8s component autodetection.

i don't think the k8s project should document vbox options.
having it in blog posts seems fine.

In this post I will look at how to make all of the changes needed for
a single-control-node cluster.  I have not yet attempted multiple
control nodes, and the solution will need further work for that case
(to give each kube-apiserver a distinct advertise-address).  The
needed results are getting the desired address to appear in the

> following places.

Author: As far as I know, you have recited facts about Kubernetes.  If Kubernetes has an simple automated way to handle these scenarios,  please give a pointer.

> following places.

Reviewer 3: i don't think there is an automated way to achieve this.

the kubeadm "create a cluster" page has the following:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/

> (Optional) Unless otherwise specified, kubeadm uses the network interface associated with the default gateway to set the advertise address for this particular control-plane node's API server. To use a different network interface, specify the --apiserver-advertise-address=<ip-address> argument to kubeadm init. To deploy an IPv6 Kubernetes cluster using IPv6 addressing, you must specify an IPv6 address, for example --apiserver-advertise-address=fd00::101

we could add another note that indicates to users that a similar IP detection is done in all components. relevant flags of said components, already document this. if the user interface setup is "non-compliant", they would have to override a number of flags rendering this setup as not-recommended.

command line.  You can get `kubeadm init` to do that either with a
command line flag or a bit of configuration file content.  The command
line flag would be `--apiserver-advertise-address=192.168.56.101`,
assuming the desired address for the control node is 192.168.56.101.
I will use that example address throughout this post.

I used a configuration file, named `ka-init.conf`.  The following
content will get this job done.

```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
apiServer:
  extraArgs:
    advertise-address: 192.168.56.101
```

## The Subject Alternative Name list in the kube-apiserver's certificate

This job can be done with either a `kubeadm init` command line flag or
a bit of config file contents.  The command line flag is
`--apiserver-cert-extra-sans=192.168.56.101`.  The config file content
sets the `apiServer.certSANS` of a `ClusterConfiguration` to hold the
desired extra entries, such as `IP:192.168.56.101`.  Because the
Subject Alternative Name list is a list, it is not functionally
critical to make that list exclude the undesired default address.  If
your use case places a premium on security, you will want to think

> about also removing the irrelevant address from that list.

Author: Perhaps I had a different experience because I used a ClusterConfiguration instead of an InitConfiguration.

> about also removing the irrelevant address from that list.

Reviewer 3: yes, in your k/k issue you mentioned that you passed ClusterConfiguration.apiServer.extraArgs which is not used by kubeadm logic, it's just passed to the component.

the advertise address flag and field in Init|JoinConfiguration are "managed" and should be used instead.

Alternative Name list.

The command line flag is `--control-plane-endpoint=192.168.56.101`.

The config file content sets the `controlPlaneEndpoint` of a
`ClusterConfiguration`.  Following is the cumulative config file
content so far.

```yaml
apiVersion: kubeadm.k8s.io/v1beta2
kind: ClusterConfiguration
apiServer:
  extraArgs:
    advertise-address: 192.168.56.101
controlPlaneEndpoint: 192.168.56.101
```

## Each node's InternalIP type of address

The `kubeadm` tool takes no responsibility for configuring this.
However, it is part of the problem at hand.  I solved it by looking
into the `systemd` configuration for the kubelet and discovering that:
- the kubelet command line includes `$KUBELET_EXTRA_ARGS`,
- in a vanilla installation, nothing sets that variable,
- the file `/etc/default/kubelet`, if it exists, is `source`d as part
  of preparing the kubelet command line,
- in a vanilla installation, that file does not exist.

So I was able to use a simple solution that does not conflict with any
maintained configuration.  On each node I create the file named
`/etc/default/kubelet` and give it content like the following:

```shell
KUBELET_EXTRA_ARGS=--node-ip=192.168.56.101

> ```

Reviewer 2: this is not true, you can configure the node-ip from kubeadm patching the kubelet configuration
```
---
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
metadata:
  name: config
cgroupDriver: cgroupfs
address: "192.168.56.101"
healthzBindAddress: "::"
```

if the address is 0.0.0.0 or empty it autodetects it, prefer IPv4, if the address is "::" autodetects and prefer IPv6

**CAUTION:**
1. The kubelet also accepts a command line flag named `--address`, but
   that one does not get this job done.

> 2. The proper content for this file is different on every node.

Author: There is also an explanation at https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/kubelet-integration/ .  This one mentions the same file in https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/kubelet-integration/#the-kubelet-drop-in-file-for-systemd , covering all the distros that use debs or RPMs.  Are there distros that do not use either of those?

The real reason I did not use `{Init|Join}Configuration.NodeRegistrationOptions.KubeletExtraArgs` is that I never even noticed JoinConfiguration --- the kubeadm doc pretty prominently exhibits `kubeadm join` commands that do not mention any config file --- and I was wading through a lot of new pages, most of which do not have links to the actual type definitions needed.

> 2. The proper content for this file is different on every node.

Reviewer 3: the page about kubeadm join page has a section about config:
https://kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-join/#config-file
it links also links to the API docs.

> ## Flannel and kubeadm

Author: I think a good blog post should be about a specific experience.  What would you recommend instead of Flannel, for someone looking for something quick and easy?

> ## Flannel and kubeadm

Reviewer 3: Calico, Cilium and WeaveNet have more positive feedback.
